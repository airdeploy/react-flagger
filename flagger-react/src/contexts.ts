import {IEntity, IFlaggerConfiguration} from 'flagger/types/Types'
import React from 'react'

export interface IFlaggerCtx {
  entity?: IEntity
  config?: IFlaggerConfiguration
  loading: boolean
  getVariation: (codename: string, entity?: IEntity) => string
  getFlagDetails: (
    codename: string,
    entity?: IEntity
  ) => {variation: string; enabled: boolean; isSampled: boolean; payload: any}
}

export const flaggerCtx = React.createContext<IFlaggerCtx>({
  loading: false,
  getVariation: () => 'off',
  getFlagDetails: () => ({
    variation: 'off',
    enabled: false,
    isSampled: false,
    payload: null,
  }),
})
