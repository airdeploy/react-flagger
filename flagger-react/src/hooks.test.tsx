import React from 'react'
import renderer from 'react-test-renderer'

import {flaggerCtx} from './contexts'
import {useFlag, useVariation} from './index'

const ViewJson = (props: any) => <div>{JSON.stringify(props, null, 2)}</div>

describe('hooks.ts tests', () => {
  beforeAll(async () => {
    jest.mock('flagger')
  })

  afterAll(async () => {
    jest.clearAllMocks()
  })

  describe('useFlag() tests', () => {
    test('uses data from flaggerCtx', () => {
      const HookComponent = () => {
        const results = useFlag('example')
        return <ViewJson {...results} />
      }

      const entity = {id: '1'}
      const getFlagDetails = jest.fn(() => {
        return {
          variation: 'some',
          enabled: true,
          isSampled: true,
          payload: {},
        }
      })
      const getVariation = jest.fn(() => 'some')

      const component = renderer.create(
        <flaggerCtx.Provider
          value={{
            entity,
            loading: false,
            getVariation,
            getFlagDetails,
          }}>
          <HookComponent />
        </flaggerCtx.Provider>
      )

      const expectedRender = renderer.create(
        <ViewJson
          loading={false}
          variation="some"
          enabled={true}
          isSampled={true}
          payload={{}}
          codename="example"
          entity={entity}
        />
      )

      expect(component.toJSON()).toMatchSnapshot()
      expect(component.toJSON()).toEqual(expectedRender.toJSON())
      expect(getFlagDetails).toBeCalledTimes(1)
      expect(getFlagDetails.mock.calls[0]).toEqual(['example', entity])
    })

    test('uses provided entity', () => {
      const entity2 = {id: '2'}
      const HookComponent = () => {
        const results = useFlag('example', entity2)
        return <ViewJson {...results} />
      }

      const entity = {id: '1'}
      const getFlagDetails = jest.fn(() => {
        return {
          variation: 'some',
          enabled: true,
          isSampled: true,
          payload: {},
        }
      })
      const getVariation = jest.fn(() => 'some')

      const component = renderer.create(
        <flaggerCtx.Provider
          value={{
            entity,
            loading: false,
            getVariation,
            getFlagDetails,
          }}>
          <HookComponent />
        </flaggerCtx.Provider>
      )

      const expectedRender = renderer.create(
        <ViewJson
          loading={false}
          variation="some"
          enabled={true}
          isSampled={true}
          payload={{}}
          codename="example"
          entity={entity2}
        />
      )

      expect(component.toJSON()).toMatchSnapshot()
      expect(component.toJSON()).toEqual(expectedRender.toJSON())
      expect(getFlagDetails).toBeCalledTimes(1)
      expect(getFlagDetails.mock.calls[0]).toEqual(['example', entity2])
    })

    test('does not recalculate results if nothing changes', () => {
      const entity2 = {id: '2'}
      const HookComponent = () => {
        const results = useFlag('example', entity2)
        return <ViewJson {...results} />
      }

      const entity = {id: '1'}
      const getFlagDetails = jest.fn(() => {
        return {
          variation: 'some',
          enabled: true,
          isSampled: true,
          payload: {},
        }
      })
      const getVariation = jest.fn(() => 'some')

      const component = renderer.create(
        <flaggerCtx.Provider
          value={{
            entity,
            loading: false,
            getVariation,
            getFlagDetails,
          }}>
          <HookComponent />
        </flaggerCtx.Provider>
      )
      component.update(
        <flaggerCtx.Provider
          value={{
            entity,
            loading: false,
            getVariation,
            getFlagDetails,
          }}>
          <HookComponent />
        </flaggerCtx.Provider>
      )

      const expectedRender = renderer.create(
        <ViewJson
          loading={false}
          variation="some"
          enabled={true}
          isSampled={true}
          payload={{}}
          codename="example"
          entity={entity2}
        />
      )

      expect(component.toJSON()).toMatchSnapshot()
      expect(component.toJSON()).toEqual(expectedRender.toJSON())
      expect(getFlagDetails).toBeCalledTimes(1)
    })

    test('recalculates results if something changes', () => {
      const entity2 = {id: '2'}
      const HookComponent = () => {
        const results = useFlag('example', entity2)
        return <ViewJson {...results} />
      }

      const entity = {id: '1'}
      const config = {
        hashKey: '',
        sdkConfig: {
          SDK_INGESTION_INTERVAL: 100,
          SDK_INGESTION_MAX_CALLS: 10,
        },
      }
      const getFlagDetails = jest.fn(() => {
        return {
          variation: 'some',
          enabled: true,
          isSampled: true,
          payload: {},
        }
      })
      const getVariation = jest.fn(() => 'some')

      const component = renderer.create(
        <flaggerCtx.Provider
          value={{
            entity,
            loading: false,
            getVariation,
            getFlagDetails,
          }}>
          <HookComponent />
        </flaggerCtx.Provider>
      )
      component.update(
        <flaggerCtx.Provider
          value={{
            entity,
            config,
            loading: false,
            getVariation,
            getFlagDetails,
          }}>
          <HookComponent />
        </flaggerCtx.Provider>
      )

      const expectedRender = renderer.create(
        <ViewJson
          loading={false}
          variation="some"
          enabled={true}
          isSampled={true}
          payload={{}}
          codename="example"
          entity={entity2}
        />
      )

      expect(component.toJSON()).toMatchSnapshot()
      expect(component.toJSON()).toEqual(expectedRender.toJSON())
      expect(getFlagDetails).toBeCalledTimes(2)
    })
  })

  describe('useVariation() tests', () => {
    test('uses data from flaggerCtx', () => {
      const HookComponent = () => {
        const results = useVariation('example')
        return <ViewJson {...results} />
      }

      const entity = {id: '1'}
      const getFlagDetails = jest.fn(() => {
        return {
          variation: 'some',
          enabled: true,
          isSampled: true,
          payload: {},
        }
      })
      const getVariation = jest.fn(() => 'hey')

      const component = renderer.create(
        <flaggerCtx.Provider
          value={{
            entity,
            loading: false,
            getVariation,
            getFlagDetails,
          }}>
          <HookComponent />
        </flaggerCtx.Provider>
      )

      const expectedRender = renderer.create(
        <ViewJson
          codename="example"
          entity={entity}
          variation="hey"
          loading={false}
        />
      )

      expect(component.toJSON()).toMatchSnapshot()
      expect(component.toJSON()).toEqual(expectedRender.toJSON())
      expect(getFlagDetails).not.toBeCalled()
      expect(getVariation).toBeCalledTimes(1)
      expect(getVariation.mock.calls[0]).toEqual(['example', entity])
    })

    test('uses provided entity', () => {
      const entity2 = {id: '2'}
      const HookComponent = () => {
        const results = useVariation('example', entity2)
        return <ViewJson {...results} />
      }

      const entity = {id: '1'}
      const getFlagDetails = jest.fn(() => {
        return {
          variation: 'some',
          enabled: true,
          isSampled: true,
          payload: {},
        }
      })
      const getVariation = jest.fn(() => 'some')

      const component = renderer.create(
        <flaggerCtx.Provider
          value={{
            entity,
            loading: false,
            getVariation,
            getFlagDetails,
          }}>
          <HookComponent />
        </flaggerCtx.Provider>
      )

      const expectedRender = renderer.create(
        <ViewJson
          codename="example"
          entity={entity2}
          variation="some"
          loading={false}
        />
      )

      expect(component.toJSON()).toMatchSnapshot()
      expect(component.toJSON()).toEqual(expectedRender.toJSON())
      expect(getVariation).toBeCalledTimes(1)
      expect(getVariation.mock.calls[0]).toEqual(['example', entity2])
    })

    test('does not recalculate variation if nothing changes', () => {
      const entity2 = {id: '2'}
      const HookComponent = () => {
        const results = useVariation('example', entity2)
        return <ViewJson {...results} />
      }

      const entity = {id: '1'}
      const getFlagDetails = jest.fn(() => {
        return {
          variation: 'some',
          enabled: true,
          isSampled: true,
          payload: {},
        }
      })
      const getVariation = jest.fn(() => 'some')

      const component = renderer.create(
        <flaggerCtx.Provider
          value={{
            entity,
            loading: false,
            getVariation,
            getFlagDetails,
          }}>
          <HookComponent />
        </flaggerCtx.Provider>
      )
      component.update(
        <flaggerCtx.Provider
          value={{
            entity,
            loading: false,
            getVariation,
            getFlagDetails,
          }}>
          <HookComponent />
        </flaggerCtx.Provider>
      )

      const expectedRender = renderer.create(
        <ViewJson
          codename="example"
          entity={entity2}
          variation="some"
          loading={false}
        />
      )

      expect(component.toJSON()).toMatchSnapshot()
      expect(component.toJSON()).toEqual(expectedRender.toJSON())
      expect(getVariation).toBeCalledTimes(1)
      expect(getVariation.mock.calls[0]).toEqual(['example', entity2])
    })

    test('recalculates variation if something changes', () => {
      const entity2 = {id: '2'}
      const HookComponent = () => {
        const results = useVariation('example', entity2)
        return <ViewJson {...results} />
      }

      const entity = {id: '1'}
      const config = {
        hashKey: '',
        sdkConfig: {
          SDK_INGESTION_INTERVAL: 100,
          SDK_INGESTION_MAX_CALLS: 10,
        },
      }
      const getFlagDetails = jest.fn(() => {
        return {
          variation: 'some',
          enabled: true,
          isSampled: true,
          payload: {},
        }
      })
      const getVariation = jest.fn(() => 'some')

      const component = renderer.create(
        <flaggerCtx.Provider
          value={{
            entity,
            loading: false,
            getVariation,
            getFlagDetails,
          }}>
          <HookComponent />
        </flaggerCtx.Provider>
      )
      component.update(
        <flaggerCtx.Provider
          value={{
            entity,
            config,
            loading: false,
            getVariation,
            getFlagDetails,
          }}>
          <HookComponent />
        </flaggerCtx.Provider>
      )

      const expectedRender = renderer.create(
        <ViewJson
          codename="example"
          entity={entity2}
          variation="some"
          loading={false}
        />
      )

      expect(component.toJSON()).toMatchSnapshot()
      expect(component.toJSON()).toEqual(expectedRender.toJSON())
      expect(getVariation).toBeCalledTimes(2)
    })
  })
})
