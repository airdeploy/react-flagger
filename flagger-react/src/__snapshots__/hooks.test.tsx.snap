// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`hooks.ts tests useFlag() tests does not recalculate results if nothing changes 1`] = `
<div>
  {
  "loading": false,
  "variation": "some",
  "enabled": true,
  "isSampled": true,
  "payload": {},
  "codename": "example",
  "entity": {
    "id": "2"
  }
}
</div>
`;

exports[`hooks.ts tests useFlag() tests recalculates results if something changes 1`] = `
<div>
  {
  "loading": false,
  "variation": "some",
  "enabled": true,
  "isSampled": true,
  "payload": {},
  "codename": "example",
  "entity": {
    "id": "2"
  }
}
</div>
`;

exports[`hooks.ts tests useFlag() tests uses data from flaggerCtx 1`] = `
<div>
  {
  "loading": false,
  "variation": "some",
  "enabled": true,
  "isSampled": true,
  "payload": {},
  "codename": "example",
  "entity": {
    "id": "1"
  }
}
</div>
`;

exports[`hooks.ts tests useFlag() tests uses provided entity 1`] = `
<div>
  {
  "loading": false,
  "variation": "some",
  "enabled": true,
  "isSampled": true,
  "payload": {},
  "codename": "example",
  "entity": {
    "id": "2"
  }
}
</div>
`;

exports[`hooks.ts tests useVariation() tests does not recalculate variation if nothing changes 1`] = `
<div>
  {
  "codename": "example",
  "entity": {
    "id": "2"
  },
  "variation": "some",
  "loading": false
}
</div>
`;

exports[`hooks.ts tests useVariation() tests recalculates variation if something changes 1`] = `
<div>
  {
  "codename": "example",
  "entity": {
    "id": "2"
  },
  "variation": "some",
  "loading": false
}
</div>
`;

exports[`hooks.ts tests useVariation() tests uses data from flaggerCtx 1`] = `
<div>
  {
  "codename": "example",
  "entity": {
    "id": "1"
  },
  "variation": "hey",
  "loading": false
}
</div>
`;

exports[`hooks.ts tests useVariation() tests uses provided entity 1`] = `
<div>
  {
  "codename": "example",
  "entity": {
    "id": "2"
  },
  "variation": "some",
  "loading": false
}
</div>
`;
